explorer/src/providers/accounts/rewards.tsx

```diff
@@ -0,0 +1,201 @@
+  rewards: (InflationReward | null)[];
+  foundOldest?: boolean;
+};
+
+export type RewardsUpdate = {
+  rewards: (InflationReward | null)[];
+  foundOldest?: boolean;
+};
+
+type State = Cache.State<Rewards>;
+type Dispatch = Cache.Dispatch<RewardsUpdate>;
+
+function reconcile(
+  rewards: Rewards | undefined,
+  update: RewardsUpdate | undefined
+): Rewards | undefined {
+  if (update?.rewards === undefined) {
+    return rewards;
+  }
+
+  const combined = (rewards?.rewards || [])
+    .concat(update.rewards)
+    .filter((value) => value !== null);
+
+  const foundOldest =
+    update.foundOldest ||
+    update.rewards.length < 1 ||
+    update.rewards[update.rewards.length - 1]?.epoch ===
+      combined[combined.length]?.epoch;
+
+  return {
+    rewards: combined,
+    highestFetchedEpoch: combined[0]?.epoch,
+    lowestFetchedEpoch: combined[combined.length - 1]?.epoch,
+    foundOldest,
+  };
+}
+
+export const StateContext = React.createContext<State | undefined>(undefined);
+export const DispatchContext = React.createContext<Dispatch | undefined>(
+  undefined
+);
+
+type RewardsProviderProps = { children: React.ReactNode };
+
+export function RewardsProvider({ children }: RewardsProviderProps) {
+  const { url } = useCluster();
+  const [state, dispatch] = Cache.useCustomReducer(url, reconcile);
+
+  React.useEffect(() => {
+    dispatch({ type: ActionType.Clear, url });
+  }, [dispatch, url]);
+
+  return (
+    <StateContext.Provider value={state}>
+      <DispatchContext.Provider value={dispatch}>
+        {children}
+      </DispatchContext.Provider>
+    </StateContext.Provider>
+  );
+}
+
+async function fetchRewards(
+  dispatch: Dispatch,
+  pubkey: PublicKey,
+  cluster: Cluster,
+  url: string,
+  fromEpoch?: number,
+  highestEpoch?: number,
+  lowestEpoch?: number
+) {
+  dispatch({
+    type: ActionType.Update,
+    status: FetchStatus.Fetching,
+    key: pubkey.toBase58(),
+    url,
+  });
+
+  const connection = new Connection(url);
+
+  if (!fromEpoch && highestEpoch) {
+    fromEpoch = highestEpoch;
+  }
+
+  if (!fromEpoch) {
+    try {
+      const epochInfo = await connection.getEpochInfo();
+      fromEpoch = epochInfo.epoch - 1;
+    } catch (error) {
+      if (cluster !== Cluster.Custom) {
+        reportError(error, { url });
+      }
+
+      return dispatch({
+        type: ActionType.Update,
+        status: FetchStatus.FetchFailed,
+        key: pubkey.toBase58(),
+        url,
+      });
+    }
+  }
+
+  const getInflationReward = async (epoch: number) => {
+    try {
+      const result = await connection.getInflationReward([pubkey], epoch);
+      return result[0];
+    } catch (error) {
+      if (cluster !== Cluster.Custom) {
+        reportError(error, { url });
+      }
+    }
+    return null;
+  };
+
+  lowestEpoch = lowestEpoch || 0;
+  const requests = [];
+  for (let i = fromEpoch; i > fromEpoch - PAGE_SIZE; i--) {
+    if (i >= lowestEpoch) {
+      requests.push(getInflationReward(i));
```

Comment:
We really need a way to batch rpc's in a generic way, eh?

Reply:
Yes, I've been thinking about this...